cpp/src/decisiontree/quantile/quantile.cuh

```diff
@@ -183,5 +183,70 @@ void preprocess_quantile(const T *data, const unsigned int *rowids,
   return;
 }
 
+template <typename T>
+__global__ void computeQuantilesSorted(T *quantiles, const int n_bins,
+                                       const T *sorted_data, const int length) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  float bin_width = float(length) / n_bins;
+  int index = int(round((tid + 1) * bin_width)) - 1;
+  // Old way of computing quantiles. Kept here for comparison.
+  // To be deleted eventually
+  // int index = (tid + 1) * floor(bin_width) - 1;
+  if (tid < n_bins) {
+    quantiles[tid] = sorted_data[index];
+  }
+
+  return;
+}
+
+template <typename T>
+void computeQuantiles(
+  T *quantiles, int n_bins, const T *data, int n_rows, int n_cols,
+  const std::shared_ptr<MLCommon::deviceAllocator> device_allocator,
+  cudaStream_t stream) {
+  // Determine temporary device storage requirements
+  MLCommon::device_buffer<char> *d_temp_storage = nullptr;
+  size_t temp_storage_bytes = 0;
+
+  MLCommon::device_buffer<T> *single_column_sorted;
+  single_column_sorted =
+    new MLCommon::device_buffer<T>(device_allocator, stream, n_rows);
+
+  CUDA_CHECK(cub::DeviceRadixSort::SortKeys(d_temp_storage, temp_storage_bytes,
+                                            data, single_column_sorted->data(),
+                                            n_rows, 0, 8 * sizeof(T), stream));
+
+  // Allocate temporary storage for sorting
+  d_temp_storage = new MLCommon::device_buffer<char>(device_allocator, stream,
+                                                     temp_storage_bytes);
+
+  // Compute quantiles column by column
+  for (int col = 0; col < n_cols; col++) {
+    int col_offset = col * n_rows;
+    int quantile_offset = col * n_bins;
+
+    CUDA_CHECK(cub::DeviceRadixSort::SortKeys(
+      (void *)d_temp_storage->data(), temp_storage_bytes, &data[col_offset],
+      single_column_sorted->data(), n_rows, 0, 8 * sizeof(T), stream));
+
+    int blocks = raft::ceildiv(n_bins, 128);
+
+    computeQuantilesSorted<<<blocks, 128, 0, stream>>>(
+      &quantiles[quantile_offset], n_bins, single_column_sorted->data(),
+      n_rows);
+
+    CUDA_CHECK(cudaGetLastError());
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```

Comment:
any reason we need the stream-sync here?

Reply:
I think this left over code added while debugging. I'll remove it now. Thanks for catching.